Release-v2.0.0

rschouten97 opened the pull request:
Changed to use of Access Token Authentication instead of Certificate Based Authentication
Removed building an additional PS session to hold the EXO session
Added check if all mandatory values from the configuration object are available
Added check if the correlation fields are available
Added check if all mandatory values from the account object are available
Updated logging to latest best practices

rhouthuijzen reviewed (commented):
An App Registration in Azure AD is required. Please follow the Microsoft documentation as reference to configure the App Registration correctly

The link is based on a certificate

rhouthuijzen reviewed (commented):
Microsoft Exchange Online PowerShell V2 module 

Change to V3.1 (multiple times in readme)

rhouthuijzen commented on create.correlateUpdateMailboxFolderPermission.ps1:
```diff
@@ -59,500 +67,264 @@ function Resolve-HTTPError {
+        Write-Verbose "Updating folder permissions for mailbox [$($aRef.userPrincipalName) ($($aRef.Guid))]: $($mailboxSplatParams | ConvertTo-Json)"
+
+        # Get Mailbox "Calendar" folder name
+        $mailboxFolderId = Get-MailboxFolderStatistics -Identity $mailbox.Guid -FolderScope Calendar | Where-Object { $_.FolderType -eq 'Calendar' } | Select-Object Name
```
I get this result: @{Name=Agenda}

Suggestion:
(Get-MailboxFolderStatistics -Identity $mailbox.Guid -FolderScope Calendar | Where-Object { $_.FolderType -eq 'Calendar' } | Select-Object Name).Name

Result: Agenda

rschouten97 replied:
Good catch, corrected.

rhouthuijzen commented on create.correlateUpdateMailboxRegionalConfigurationAndFolderPermission.ps1:
```diff
@@ -88,630 +74,303 @@ function Resolve-HTTPError {
+        Write-Verbose "Updating folder permissions for mailbox [$($aRef.userPrincipalName) ($($aRef.Guid))]: $($mailboxSplatParams | ConvertTo-Json)"
+
+        # Get Mailbox "Calendar" folder name
+        $mailboxFolderId = Get-MailboxFolderStatistics -Identity $mailbox.Guid -FolderScope Calendar | Where-Object { $_.FolderType -eq 'Calendar' } | Select-Object Name
```
I get this result: @{Name=Agenda}

Suggestion:
(Get-MailboxFolderStatistics -Identity $mailbox.Guid -FolderScope Calendar | Where-Object { $_.FolderType -eq 'Calendar' } | Select-Object Name).Name

Result: Agenda

rschouten97 replied:
Good catch. corrected

rhouthuijzen commented on delete.UpdateMailboxAutoReplyConfiguration.ps1:
```diff
@@ -78,498 +68,243 @@ function Resolve-HTTPError {
+        Write-Verbose "Updating autoreply configuration for mailbox [$($aRef.userPrincipalName) ($($aRef.Guid))]: $($mailboxSplatParams | ConvertTo-Json)"
+
+        # Get Mailbox "Calendar" folder name
+        $mailboxFolderId = Get-MailboxFolderStatistics -Identity $mailbox.Guid -FolderScope Calendar | Where-Object { $_.FolderType -eq 'Calendar' } | Select-Object Name
```
$mailboxFolderId is not used. So I think the Get-MailboxFolderStatistics can be removed

rschouten97 replied:
Removed

rhouthuijzen commented on revoke.sharedmailboxes.ps1:
```diff
@@ -69,576 +55,331 @@ function Resolve-HTTPError {
+                    } 
+
+                    if ($dryRun -eq $false) {
+                        $addFullAccessPermission = Add-MailboxPermission @FullAccessPermissionSplatParams -ErrorAction Stop
```
Add-MailboxPermission --> Remove-MailboxPermission

rschouten97 replied:
Corrected

rhouthuijzen commented on revoke.sharedmailboxes.ps1:
```diff
@@ -69,576 +55,331 @@ function Resolve-HTTPError {
+                try {
+                    Write-Verbose "Revoking permission [FullAccess] to mailbox [$($pRef.Name) ($($pRef.id))] for user [$($aRef.UserPrincipalName) ($($aRef.Guid))]"
+
+                    $FullAccessPermissionSplatParams = @{
```
Update SplatParms to:

                    Identity        = $pRef.id
                    User            = $aRef.Guid
                    AccessRights    = 'FullAccess'
                    InheritanceType = 'All'
                    Confirm         = $false

rschouten97 replied:
Updated

rhouthuijzen commented on revoke.sharedmailboxes.ps1:
```diff
@@ -69,576 +55,331 @@ function Resolve-HTTPError {
+        $authUri = $baseUri + "$AADTenantId/oauth2/token"
+        
+        $body = @{
+            revoke_type   = "client_credentials"
```
revoke_type --> grant_type

rhouthuijzen commented on dynamicPermission.SharedMailboxes.ReferenceExample.ps1:
```diff
@@ -0,0 +1,735 @@
+                } 
+
+                if ($dryRun -eq $false) {
+                    $addFullAccessPermission = Add-MailboxPermission @FullAccessPermissionSplatParams -ErrorAction Stop
```
Not tested. But Add-MailboxPermissions seems odd when revoking permissions

rschouten97 replied:
corrected

rhouthuijzen commented on dynamicPermission.SharedMailboxes.ReferenceExample.ps1:
```diff
@@ -0,0 +1,735 @@
+                } 
+
+                if ($dryRun -eq $false) {
+                    $addSendAsPermission = Add-RecipientPermission @sendAsPermissionSplatParams -ErrorAction Stop
```
Not tested. But Add-RecipientPermission seems odd when revoking permissions

rhouthuijzen commented on revoke.sharedmailboxes.ps1:
```diff
@@ -21,39 +17,29 @@ $auditLogs = [System.Collections.Generic.List[PSCustomObject]]::new()
+# Define configuration properties as required
+$requiredConfigurationFields = @("AzureADOrganization", "AzureADTenantId", "AzureADAppId", "AzureADAppSecret")
+
 # Used to connect to Exchange Online in an unattended scripting scenario using a certificate.
```
Certifcate is nolonger used

rschouten97 replied:
corrected.

rhouthuijzen commented on revoke.groups.ps1:
```diff
@@ -21,39 +17,28 @@ $auditLogs = [System.Collections.Generic.List[PSCustomObject]]::new()
+# Define configuration properties as required
+$requiredConfigurationFields = @("AzureADOrganization", "AzureADTenantId", "AzureADAppId", "AzureADAppSecret")
+
 # Used to connect to Exchange Online in an unattended scripting scenario using a certificate.
```
Certifcate is nolonger used

rschouten97 replied:
corrected.

rschouten97 commented:
> An App Registration in Azure AD is required. Please follow the Microsoft documentation as reference to configure the App Registration correctly
> 
> The link is based on a certificate



> An App Registration in Azure AD is required. Please follow the Microsoft documentation as reference to configure the App Registration correctly
> 
> The link is based on a certificate

The official Microsoft Documentation still only references the use of Certificate Based Authentication.
I removed the notes that you should follow the microsoft docs and added notes that our setup differs by using the access token.

rschouten97 commented:
> Microsoft Exchange Online PowerShell V2 module
> 
> Change to V3.1 (multiple times in readme)

Changed all references from v2 to v3.1. However, the microsoft docs articles still have the v2 url, I left this unchanged of course.

rhouthuijzen commented on revoke.sharedmailboxes.ps1:
```diff
@@ -69,576 +54,330 @@ function Resolve-HTTPError {
+        $authUri = $baseUri + "$AADTenantId/oauth2/token"
+        
+        $body = @{
+            revoke_type   = "client_credentials"
```
This should be grant_type

rschouten97 replied:
Corrected

rhouthuijzen commented:
Possibly change Azure AD role:

Assign Azure AD roles to the application
Azure AD has more than 50 admin roles available. The Global Administrator

The Global Administrator Rol is not required for the following ExO functions (tested):

    "Get-EXOMailbox"
    "Set-MailboxFolderPermission"
    "Get-MailboxFolderStatistics"
    "Add-MailboxFolderPermission"
    "Remove-MailboxFolderPermission"
    "Remove-MailboxPermission"
    "Remove-RecipientPermission"
    "Get-User"
    "Add-MailboxPermission"
    "Add-RecipientPermission"
    "Set-Mailbox"

rhouthuijzen reviewed (approved):
Approved. Used multiple times in implementations.
